README.md
# bold_identification

## 1 Introduction

see `https://github.com/linzhi2013/bold_identification`.

This is a Python3 package which can get the taxonomy information of sequences from BOLD [http://www.boldsystems.org/index.php](http://www.boldsystems.org/index.php).

Currently, `bold_identification` only runs on Mac OS, Windows 64bit, Linux.


## 2 Installation 

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/bold-identification/README.html)

Or with `pip`

    $ pip install bold_identification

There will be a command `bold_identification` created under the same directory as your `pip` command.

## 3 Usage
run `bold_identification`

    $ bold_identification
    usage: bold_identification [-h] -i <str> [-f <str>] -o <str>
                                [-d {COX1,COX1_SPECIES,COX1_SPECIES_PUBLIC,COX1_L640bp,ITS,MATK_RBCL}] [-n <int>] [-r <int>]
                                [-c] [-C] [-q <int>] [-D] [--version]

    To identify taxa of given sequences from BOLD (http://www.boldsystems.org/).
    Some sequences can fail to get taxon information, which can be caused by
    TimeoutException if your network to the BOLD server is bad.
    Those sequences will be output in the file '*.TimeoutException.fasta'.

    You can:
    1) run another searching with the same command directly (but add -c option);
    2) lengthen the time to wait for each query (-t option);
    3) increase submission times (-r option) for a sequence.

    Also, the sequences without BOLD matches will be output in the
    file '*.NoBoldMatchError.fasta'

    By Guanliang Meng.
    See https://github.com/linzhi2013/bold_identification.

    Citation:
    Yang C, Zheng Y, Tan S, Meng G, et al.
    Efficient COI barcoding using high throughput single-end 400 bp sequencing.
    https://doi.org/10.1186/s12864-020-07255-w

    version: 0.0.27

    optional arguments:
      -h, --help            show this help message and exit
      -i <str>              input file name
      -f <str>              input file format [fasta]
      -o <str>              outfile prefix
      -d {COX1,COX1_SPECIES,COX1_SPECIES_PUBLIC,COX1_L640bp,ITS,MATK_RBCL}
                            database to search [COX1]
      -n <int>              how many first top hits will be output. [1]
      -r <int>              Maximum submission time for a sequence, useful for handling TimeOutException. [4]
      -c                    continuous mode, jump over the ones already in "-o" file, will resubmit all the remained. use "-cc"
                            to also jump over the ones in "*.NoBoldMatchError.fasta" file.
      -C                    For chimera check purpose. If set, for each sequence, I will query the BOLD database using the
                            subsequences from 5'- and 3'-ends with a length of '-q <int>' bp, respectively
      -q <int>              The length of subsequences for chimera check [400]
      -D                    debug mode output [False]
      --version             show program's version number and exit



## 4 Citation
When you use `bold_identification` in your study, please cite:

    Yang C, Zheng Y, Tan S, Meng G, et al.
    Efficient COI barcoding using high throughput single-end 400 bp sequencing.
    https://doi.org/10.1186/s12864-020-07255-w











changelog

v0.0.27
20210107
Fixed a bug when search COX1_SPECIES etc. database for the result class of HMTL.


v0.0.26
20190708
Add chimera check support.


v0.0.25
20190331
use 'html5lib' to deal page source, to get best results


v0.0.22
20190331
use requests + bs4 packages to replace seleum

v0.0.20
20181121
change biopython verion requirement to be >1.5


v0.0.20
20181028
remove the empty lines from the output "-o".
fixed a logging error.
Add more NoBoldMatchError.

v0.0.19
20181027
fixed a bug, '-n 10' couldn't output the top 10 hits, only the first could be output.


v0.0.18
20181026
Add PhantomJS browser support, which does not need an graphical window natively.

v0.0.17
20181026
Add '-H' mode, i.e., no graphical window. But only Chrome succeeded in Mac and Ubuntu, while Firefox failed. Windows system has not been tested on.


v0.0.16
20181024
Add '-c' mode.
Add '-debug' option.
Add '-f' option, to specify input file format.
Add can deal with NoBoldMatchError now.

v0.0.15
20181024
resubmission of v0.0.14.

v0.0.14
20181023

Now the Animal pannel has been tested working with Chrome. the funig and plant pannel won't
work on this version


v0.0.13
20181023
1. reformat the script, make it meet PEP8 standard.
2. The 'TimeOutException' sequences will only go to '*.TimeoutException.fasta' file, and
 their sequences ids will not be presented on the result file (-o option) any more.
3. fixed a bug when use '-d COX1_SPECIES', e.g., always got TimeoutException.

v0.0.12
20180906
Add '-r' option, to specify the maximum time a sequence will be submited, to handle TimeOutException.

v0.0.11
20180713
Add '-n' option, users can specify how many top-hits to be output.

v0.0.8
20180711
Using logging module, and specify whether one sequence gets no result because
of TimeOutException.

v0.0.7
20180710
do not capture too exactly, to the 'td' level is fine,
but do not capture 'h' or 'em' level. Because BOLD result page won't
have 'h' or 'em' item when the result is empty.


bold_identification/logger.py
import logging
import os
import sys

def get_logger(debug=False):
    # 级别排序:CRITICAL > ERROR > WARNING > INFO > DEBUG
    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s")

    logger = logging.getLogger(__name__)
    logger.setLevel(logging.DEBUG)  # must be DEBUG, then 'ch' below works.
    # logger.setFormatter(formatter)

    fh = logging.FileHandler(os.path.basename(sys.argv[0]) + '.log')
    if debug:
        fh.setLevel(logging.DEBUG)
    else:
        fh.setLevel(logging.INFO)  # INFO level goes to the log file
    fh.setFormatter(formatter)
    logger.addHandler(fh)

    ch = logging.StreamHandler()
    if debug:
        ch.setLevel(logging.DEBUG)
    else:
        ch.setLevel(logging.WARNING)  # only WARNING level will output on screen
    ch.setFormatter(formatter)
    logger.addHandler(ch)

    return logger

bold_identification/bold_engin.py
import re
import requests
from bs4 import BeautifulSoup
import sys
import collections

# Copyright Guanliang MENG 2018-2020
# License: GPL v3

class NoBoldMatchError(Exception):
    def __init__(self, arg):
        super(NoBoldMatchError, self).__init__()
        self.arg = arg

    def __str__(self):
        return "NoBoldMatchError:\n{0}\n".format(self.arg)


class BOLD(object):
    """docstring for BOLD"""

    def __init__(self, db='COX1', seqid=None, seq=None, timeout=None):
        super(BOLD, self).__init__()
        self.db = db
        self.pane_type = self.db_to_paneType(db=self.db)
        self.url = self.paneType_to_url(pane_type=self.pane_type)
        self.seqid = seqid
        self.seq = re.sub(r'\s+', '', seq)
        self.timeout = timeout

        self.data = {
             'tabtype': self.pane_type,
             'searchdb': self.db,
             'sequence': self.seq,
        }


    def submit(self):
        r = requests.post(self.url, data=self.data)
        result_url = 'http://www.boldsystems.org' + r.text.split('<span style="text-decoration: none" result="')[1].split('">')[0]

        page_source = requests.get(result_url).text

        return taxonRanks(
            page_source=page_source,
            seqid=self.seqid,
            pane_type=self.pane_type,
            db=self.db)


    def db_to_paneType(self, db=None):
        db_paneType = {
            'COX1': 'animalTabPane',
            'COX1_SPECIES': 'animalTabPane',
            'COX1_SPECIES_PUBLIC': 'animalTabPane',
            'COX1_L640bp': 'animalTabPane',
            'ITS': 'fungiTabPane',
            'MATK_RBCL': 'plantTabPane'}

        return db_paneType[db]

    def paneType_to_url(self, pane_type=None):
        paneType_url = {
            'animalTabPane': 'http://www.boldsystems.org/index.php/IDS_IdentificationRequest',
            'fungiTabPane':'http://www.boldsystems.org/index.php/IDS_BlastRequest',
            'plantTabPane':'http://www.boldsystems.org/index.php/IDS_BlastRequest'
        }

        return paneType_url[pane_type]


class taxonRanks(object):
    """docstring for taxonRanks"""
    def __init__(self, page_source=None, seqid=None, pane_type=None, db=None):
        super(taxonRanks, self).__init__()
        self.page_source = page_source
        self.seqid = seqid
        self.pane_type = pane_type
        self.db = db
        self.taxa = self.get_taxonRank(page_source=self.page_source, seqid=self.seqid, pane_type=self.pane_type, db=self.db)

    def get_taxonRank(self, page_source=None, seqid=None, pane_type=None, db=None):
        if 'Unable to match any records in the selected database.' in page_source:
            raise NoBoldMatchError(seqid)

        soup = BeautifulSoup(page_source, "html5lib")

        result_class = self.get_table_class(pane_type=pane_type, db=db)
        table = soup.find_all('table', class_=result_class)[0]

        head_line = True
        ranks = []
        taxa = []
        for row in table.find_all('tr'):
            if head_line:
                for col in row.find_all('td'):
                    ranks.append(col.text)
                head_line = False
                continue
            else:
                taxon_oneline = collections.OrderedDict()
                taxon_oneline['seqid'] = seqid
                for col_name, col_val in zip(ranks, row.find_all('td')):
                    taxon_oneline[col_name] = col_val.text.strip()
                taxa.append(taxon_oneline)


        return taxa

    def get_table_class(self, pane_type=None, db='COX1'):
        paneType_class = {
            'animalTabPane': {
                'COX1': 'resultsTable noborder',
                'COX1_SPECIES': 'table resultsTable noborder',
                'COX1_SPECIES_PUBLIC': 'table resultsTable noborder',
                'COX1_L640bp': 'table resultsTable noborder',},
            'fungiTabPane': {
                'ITS': 'table resultTable noborder'},
            'plantTabPane': {
                'MATK_RBCL':'table resultTable noborder'},
        }

        return paneType_class[pane_type][db]


# 'COX1', 'COX1_SPECIES', 'COX1_SPECIES_PUBLIC', 'COX1_L640bp','ITS', 'MATK_RBCL'














bold_identification/BOLDv4_identification.py
#!/usr/bin/env python3
# Copyright Guanliang MENG 2018-2019
# License: GPL v3
import re
import sys
import time
import os
import argparse
from Bio import SeqIO
from bold_engin import BOLD, NoBoldMatchError
from logger import get_logger


# see http://docs.python-requests.org/en/master/user/quickstart/

def get_parameters():
    version = '0.0.22'
    description = '''
To identify taxa of given sequences from BOLD (http://www.boldsystems.org/).
Some sequences can fail to get taxon information, which can be caused by
TimeoutException if your network to the BOLD server is bad.
Those sequences will be output in the file '*.TimeoutException.fasta'.

You can:
1) run another searching with the same command directly (but add -c option);
2) lengthen the time to wait for each query (-t option);
3) increase submission times (-r option) for a sequence.

Also, the sequences without BOLD matches will be output in the
file '*.NoBoldMatchError.fasta'

By mengguanliang AT genomics DOT cn.
See https://github.com/linzhi2013/bold_identification.

version: {version}
'''.format(version=version)

    parser = argparse.ArgumentParser(description=description,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument('-i', dest='infile', required=True, metavar='<str>',
                        help='input file name')

    parser.add_argument('-f', dest='infileformat', default='fasta',
        metavar='<str>', help='input file format [%(default)s]')

    parser.add_argument('-o', dest='outfile', required=True, metavar='<str>',
                        help='outfile')

    parser.add_argument('-d', dest='db', choices=['COX1', 'COX1_SPECIES', 'COX1_SPECIES_PUBLIC', 'COX1_L640bp','ITS', 'MATK_RBCL'],
        required=False, default='COX1',
        help='database to search [%(default)s]')

    parser.add_argument('-n', dest='topnum', type=int, default=1,
        metavar='<int>',
        help='how many first top hits will be output. [%(default)s]')

    parser.add_argument('-r', dest='submissiontimes', type=int, required=False,
        default=4, metavar='<int>',
        help='Maximum submission time for a sequence, useful for handling TimeOutException. [%(default)s]')

    parser.add_argument('-c', dest='to_continue', action='count',
        default=False, help='continuous mode, jump over the ones already in "-o" file, will resubmit all the remained. use "-cc" to also jump over the ones in "*.NoBoldMatchError.fasta" file. [%(default)s]')

    parser.add_argument('-D', dest='debug', action='store_true', default=False,
        help='debug mode output [%(default)s]')


    parser.add_argument('--version', action='version', version='%(prog)s {}'.format(version))

    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit()
    else:
        args = parser.parse_args()
        return args


def get_finished_seqids(file=None, f_format='tab'):
    seqids = set()
    if f_format == 'tab':
        with open(file, 'r') as fh:
            for i in fh:
                i = i.rstrip()
                seqid = i.split()[0]
                seqids.add(seqid)
    elif f_format == 'fasta':
        with open(file, 'r') as fh:
            for i in fh:
                i = i.strip()
                if i.startswith('>'):
                    seqid = i.split()[0].replace('>', '')
                    seqids.add(seqid)
    else:
        sys.exit('unknown infile format: {}'.format(f_format))

    return seqids


def main():
    args = get_parameters()

    logger = get_logger(debug=args.debug)

    logger.info('args: \n{0}\n'.format(args))

    # check if in continuous mode
    # and set taxa output file
    header_line = True
    finished_seqids = set()
    if args.to_continue >= 1 and os.path.exists(args.outfile):
        finished_seqids = get_finished_seqids(args.outfile, 'tab')
        fhout = open(args.outfile, 'a')
        header_line = False
    else:
        fhout = open(args.outfile, 'w')


    # set timeout and no bold match out files
    TimeoutException_file = os.path.basename(args.outfile) + '.TimeoutException.fasta'
    fhout_timeoutSeq = open(TimeoutException_file, 'w')


    NoBoldMatchError_file = os.path.basename(
        args.outfile) + '.NoBoldMatchError.fasta'
    if args.to_continue >= 2 and os.path.exists(NoBoldMatchError_file):
        finished_seqids.update(
            get_finished_seqids(NoBoldMatchError_file, 'fasta'))
        fhout_noBoldMatch = open(NoBoldMatchError_file, 'a')
    else:
        fhout_noBoldMatch = open(NoBoldMatchError_file, 'w')


    # begin searching sequences
    count = 0
    for rec in SeqIO.parse(args.infile, args.infileformat):
        count += 1

        # continuous mode
        if args.to_continue >= 1 and (rec.id in finished_seqids):
            logger.info('skip over the {0} sequence: {1}\n'.format(count, rec.id))
            continue
        logger.info('start to search the {0} sequence: {1}\n'.format(count, rec.id))

        seq = str(rec.seq)

        for i in range(1, args.submissiontimes+1):
            if i >= 2:
                logger.info('The {0} time for the submission of: {1}\t{2}\n'.format(i, count, rec.id))
            try:
                taxa = BOLD(db=args.db, seqid=rec.id, seq=seq).submit().taxa

            except NoBoldMatchError as e:
                logger.debug(e)
                SeqIO.write(rec, fhout_noBoldMatch, args.infileformat)
                # no need to re-submit
                break
            else:
                item_tot = len(taxa)
                if item_tot > 0:
                    output_num = args.topnum
                    if item_tot < args.topnum:
                        output_num = item_tot
                    for j in range(output_num):
                        item = taxa[j]
                        if count == 1 and header_line:
                            print('\t'.join(item.keys()), file=fhout)
                            header_line = False

                        line = [item[k] for k in item.keys()]
                        line = '\t'.join(line)
                        print(rec.description, line,
                            sep='\t', file=fhout,flush=True)
                    break
                else:
                    # can be Timeout, will re-submit
                    pass
        else:
            SeqIO.write(rec, fhout_timeoutSeq, args.infileformat)

        # to avoid your IP address being blocked by BOLD server.
        time.sleep(2)

    fhout.close()
    fhout_timeoutSeq.close()
    fhout_noBoldMatch.close()


if __name__ == '__main__':
    main()


bold_identification/BOLD_identification.py
#!/usr/bin/env python3
# Copyright Guanliang MENG 2018-2020
# License: GPL v3

import re
import sys
import time
import os
import argparse
from Bio import SeqIO

src_path = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, src_path)
from bold_engin import BOLD, NoBoldMatchError
from logger import get_logger


def get_parameters():
    version = '0.0.27'
    description = '''
To identify taxa of given sequences from BOLD (http://www.boldsystems.org/).
Some sequences can fail to get taxon information, which can be caused by
TimeoutException if your network to the BOLD server is bad.
Those sequences will be output in the file '*.TimeoutException.fasta'.

You can:
1) run another searching with the same command directly (but add -c option);
2) lengthen the time to wait for each query (-t option);
3) increase submission times (-r option) for a sequence.

Also, the sequences without BOLD matches will be output in the
file '*.NoBoldMatchError.fasta'

By Guanliang Meng.
See https://github.com/linzhi2013/bold_identification.

Citation:
Yang C, Zheng Y, Tan S, Meng G, et al.
Efficient COI barcoding using high throughput single-end 400 bp sequencing.
https://doi.org/10.1186/s12864-020-07255-w

version: {version}
'''.format(version=version)

    parser = argparse.ArgumentParser(description=description,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument('-i', dest='infile', required=True, metavar='<str>',
                        help='input file name')

    parser.add_argument('-f', dest='infileformat', default='fasta',
        metavar='<str>', help='input file format [%(default)s]')

    parser.add_argument('-o', dest='outprefix', required=True, metavar='<str>',
                        help='outfile prefix')

    parser.add_argument('-d', dest='db', choices=['COX1', 'COX1_SPECIES', 'COX1_SPECIES_PUBLIC', 'COX1_L640bp','ITS', 'MATK_RBCL'],
        required=False, default='COX1',
        help='database to search [%(default)s]')

    parser.add_argument('-n', dest='topnum', type=int, default=1,
        metavar='<int>',
        help='how many first top hits will be output. [%(default)s]')

    parser.add_argument('-r', dest='submissiontimes', type=int, required=False,
        default=4, metavar='<int>',
        help='Maximum submission time for a sequence, useful for handling TimeOutException. [%(default)s]')

    parser.add_argument('-c', dest='to_continue', action='count',
        default=0, help='continuous mode, jump over the ones already in "-o" file, will resubmit all the remained. use "-cc" to also jump over the ones in "*.NoBoldMatchError.fasta" file.')

    parser.add_argument('-C', dest='check_chimera', action='store_true',
        default=False,
        help="For chimera check purpose. If set, for each sequence, I will query the BOLD database using the subsequences from 5'- and 3'-ends with a length of '-q <int>' bp, respectively")

    parser.add_argument('-q', dest='chimera_len', type=int, default=400,
        metavar='<int>',
        help="The length of subsequences for chimera check [%(default)s]")

    parser.add_argument('-D', dest='debug', action='store_true', default=False,
        help='debug mode output [%(default)s]')


    parser.add_argument('--version', action='version', version='%(prog)s {}'.format(version))

    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit()
    else:
        args = parser.parse_args()
        return args


def get_finished_seqids(file=None, f_format='tab'):
    seqids = set()
    if f_format == 'tab':
        with open(file, 'r') as fh:
            for i in fh:
                i = i.rstrip()
                seqid = i.split()[0]
                seqids.add(seqid)
    elif f_format == 'fasta':
        with open(file, 'r') as fh:
            for i in fh:
                i = i.strip()
                if i.startswith('>'):
                    seqid = i.split()[0].replace('>', '')
                    seqids.add(seqid)
    else:
        sys.exit('unknown infile format: {}'.format(f_format))

    return seqids


def multi_submissions(submissiontimes=4, seq_count=0, rec=None, db='COX1', fhout=sys.stdout, fhout_noBoldMatch=sys.stderr, fhout_timeoutSeq=sys.stdout, topnum=1, header_line=False, logger=None):
    seqid = str(rec.id)
    seq = str(rec.seq)

    for i in range(1, submissiontimes+1):
        if i >= 2:
            logger.info('The {0} time for the submission of: {1}\t{2}\n'.format(i, seq_count, seqid))
        try:
            taxa = BOLD(db=db, seqid=seqid, seq=seq).submit().taxa

        except NoBoldMatchError as e:
            logger.debug(e)
            SeqIO.write(rec, fhout_noBoldMatch, 'fasta')
            # no need to re-submit
            break
        else:
            item_tot = len(taxa)
            if item_tot > 0:
                output_num = topnum
                if item_tot < topnum:
                    output_num = item_tot
                for j in range(output_num):
                    item = taxa[j]
                    if seq_count == 1 and header_line:
                        print('\t'.join(item.keys()), file=fhout)
                        header_line = False

                    line = [item[k] for k in item.keys()]
                    line = '\t'.join(line)
                    print(line, file=fhout,flush=True)
                break
            else:
                # can be Timeout, will re-submit
                continue
    else:
        SeqIO.write(rec, fhout_timeoutSeq, 'fasta')


def get_taxa(to_continue=False, infile=None, infileformat='fasta', db='COX1',  outprefix='out', submissiontimes=4, topnum=1, logger=None):
    # check if in continuous mode
    # and set taxa output file
    header_line = True
    finished_seqids = set()

    taxa_file = outprefix + '.taxa'
    if to_continue >= 1 and os.path.exists(taxa_file):
        finished_seqids = get_finished_seqids(taxa_file, 'tab')
        fhout = open(taxa_file, 'a')
        header_line = False
    else:
        fhout = open(taxa_file, 'w')


    # set timeout and no bold match out files
    fhout_timeoutSeq = open(outprefix+'.TimeoutException.fasta', 'w')


    NoBoldMatchError_file = outprefix + '.NoBoldMatchError.fasta'
    if to_continue >= 2 and os.path.exists(NoBoldMatchError_file):
        finished_seqids.update(
            get_finished_seqids(NoBoldMatchError_file, 'fasta'))
        fhout_noBoldMatch = open(NoBoldMatchError_file, 'a')
    else:
        fhout_noBoldMatch = open(NoBoldMatchError_file, 'w')


    # begin searching sequences
    seq_count = 0
    for rec in SeqIO.parse(infile, infileformat):
        seq_count += 1

        # continuous mode
        if to_continue >= 1 and (rec.id in finished_seqids):
            logger.info('skip over the {0} sequence: {1}\n'.format(seq_count, rec.id))
            continue
        logger.info('start to search the {0} sequence: {1}\n'.format(seq_count, rec.id))

        multi_submissions(
            submissiontimes=submissiontimes,
            seq_count = seq_count,
            rec = rec,
            db = db,
            fhout = fhout,
            fhout_noBoldMatch = fhout_noBoldMatch,
            fhout_timeoutSeq = fhout_timeoutSeq,
            topnum = topnum,
            header_line = header_line,
            logger = logger)

        # to avoid your IP address being blocked by BOLD server.
        time.sleep(2)

    fhout.close()
    fhout_timeoutSeq.close()
    fhout_noBoldMatch.close()


def chimera_check(infile=None, infileformat='fasta', db='COX1', chimera_len=400, outprefix='check_chimera', submissiontimes=4, topnum=1, logger=None):
    # fhout_5 = open(outprefix+'.5end.taxa', 'w')
    # fhout_3 = open(outprefix+'.3end.taxa', 'w')

    fhout = open(outprefix+'.5-and-3ends.taxa', 'w')

    fhout_noBoldMatch_5 = open(outprefix + '.5end.NoBoldMatchError.fasta', 'w')
    fhout_noBoldMatch_3 = open(outprefix + '.3end.NoBoldMatchError.fasta', 'w')

    # set timeout and no bold match out files
    fhout_timeoutSeq_5 = open(outprefix + '.5end.TimeoutException.fasta', 'w')
    fhout_timeoutSeq_3 = open(outprefix + '.3end.TimeoutException.fasta', 'w')

    seq_count = 0
    header_line_5 = True
    header_line_3 = True
    for rec in SeqIO.parse(infile, infileformat):
        seq_count += 1
        logger.info('start to search the {0} sequence: {1}\n'.format(seq_count, rec.id))

        rec_5 = rec[0:chimera_len]
        rec_5.id = rec.id + '_5end'
        rec_5.description = ''

        rec_3 = rec[-chimera_len:]
        rec_3.id = rec.id + '_3end'
        rec_3.description = ''

        multi_submissions(
            submissiontimes=submissiontimes,
            seq_count = seq_count,
            rec = rec_5,
            db = db,
            fhout = fhout,
            fhout_noBoldMatch = fhout_noBoldMatch_5,
            fhout_timeoutSeq = fhout_timeoutSeq_5,
            topnum = topnum,
            header_line = header_line_5,
            logger = logger)

        multi_submissions(
            submissiontimes=submissiontimes,
            seq_count = seq_count,
            rec = rec_3,
            db = db,
            fhout = fhout,
            fhout_noBoldMatch = fhout_noBoldMatch_3,
            fhout_timeoutSeq = fhout_timeoutSeq_3,
            topnum = topnum,
            header_line = header_line_3,
            logger = logger)

        # to avoid your IP address being blocked by BOLD server.
        time.sleep(2)

    fhout.close()
    # fhout_5.close()
    fhout_timeoutSeq_5.close()
    fhout_noBoldMatch_5.close()
    # fhout_3.close()
    fhout_timeoutSeq_3.close()
    fhout_noBoldMatch_3.close()


def main():
    args = get_parameters()

    logger = get_logger(debug=args.debug)
    logger.info('args: \n{0}\n'.format(args))

    if args.check_chimera:
        chimera_check(
            infile=args.infile,
            infileformat=args.infileformat,
            db=args.db,
            chimera_len=args.chimera_len,
            outprefix=args.outprefix,
            submissiontimes=args.submissiontimes,
            topnum=args.topnum,
            logger=logger)
    else:
        get_taxa(
            to_continue=args.to_continue,
            infile=args.infile,
            infileformat=args.infileformat,
            db=args.db,
            outprefix=args.outprefix,
            submissiontimes=args.submissiontimes,
            topnum=args.topnum,
            logger=logger)



if __name__ == '__main__':
    main()
